c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h

```diff
@@ -0,0 +1,315 @@
+	const T& GetTail() const
+	{
+		return Tail->Value;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	bool IsEmpty() const
+	{
+		return Size == 0;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	int GetSize() const
+	{
+		return Size;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void AddHead(const T& Value)
+	{
+		if (IsEmpty())
+		{
+			Head = new FNode(Value);
+			Tail = Head;
+
+			Size = 1;
+			return;
+		}
+
+		Head = new FNode(Value, Head);
+		Size++;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void AddTail(const T& Value)
+	{
+		if (IsEmpty())
+		{
+			AddHead(Value);
+			return;
+		}
+
+		Tail->Next = new FNode(Value);
+		Tail = Tail->Next;
+
+		Size++;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Insert(const T& Value, const int Index)
+	{
+		if (Index < 0 || Index > Size)
+		{
+			return;
+		}
+
+		if (Index == 0)
+		{
+			AddHead(Value);
+			return;
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			PreviousNode.Next = new FNode(Value, PreviousNode.Next);
+			
+			if (Index == Size)
+			{
+				Tail = Tail->Next;
+			}
+
+			Size++;
+		}
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Remove(const int Index)
+	{
+		if (Index < 0 || Index >= Size)
+		{
+			return;
+		}
+
+		if (Index == 0)
+		{
+			FNode* NodeToDeletePtr = Head;
+			Head = NodeToDeletePtr->Next;
+
+			delete NodeToDeletePtr;
+			NodeToDeletePtr = nullptr;
+
+			if (Size == 1)
+			{
+				Tail = nullptr;
+			}
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* NodeToDeletePtr = PreviousNode.Next;
+
+			PreviousNode.Next = NodeToDeletePtr->Next;
+
+			delete NodeToDeletePtr;
+			NodeToDeletePtr = nullptr;
+
+			if (Index == Size - 1)
+			{
+				Tail = &PreviousNode;
+			}
+		}
+
+		Size--;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		int OriginalSize = Size;
+		for (int i = 0; i < OriginalSize; i++)
+		{
+			Remove(0);
```

Comment:
No, I am deleting every element, it's just that they all become the head once its predecessor gets deleted, I also think this is the most efficient way of doing it since removing the head is the first possible instruction inside the Remove method.